### ISSUE TITLE
Internal server error produced by Siegel modular form links.

### ISSUE THREAD

**Issue by AndrewVSutherland** [LMFDB/lmfdb#1201]
I was just demoing the LMFDB and found that clicking on any of the following links (including the very first link on the main Seigel modular form page) results in an "Internal Server Error"

http://www.lmfdb.org/ModularForm/GSp/Q/Sp4Z_j/
http://www.lmfdb.org/ModularForm/GSp/Q/Gamma0_2/
http://www.lmfdb.org/ModularForm/GSp/Q/Gamma1_2/
http://www.lmfdb.org/ModularForm/GSp/Q/Gamma_2/

**Comment by sanni85** [comment]
There is a test for all the links on the browse page for Siegel modular forms. 
I wrote it this week and all tests were passing, so the link were all working.

**Comment by davidfarmer** [comment]
Works on my laptop, but it does take around 15 seconds for the
first one to render.

**Comment by sanni85** [comment]
the only one who does not work for me is http://www.lmfdb.org/ModularForm/GSp/Q/Sp4Z_j/

**Comment by AndrewVSutherland** [comment]
I just clicked on all 4 links above from (a) linux laptop, (b) windows desktop, (c) ipad mini (d) android phone and get Internal Server Error in every case.

**Comment by davidfarmer** [comment]
Still works fine (if a bit slow) on my laptop.

On www and beta, I get Internal Server Error instantly.

I have not upgraded any database software, so that may be a clue?

**Comment by AndrewVSutherland** [comment]
Hmm, is it possible we are getting redirected to different back ends?  Or that there are different worker threads being assigned?

**Comment by AndrewVSutherland** [comment]
Here is a traceroute from my machine at home:

 1  10.0.0.253 (10.0.0.253)  0.983 ms  0.979 ms  1.271 ms
 2  96.120.66.105 (96.120.66.105)  13.681 ms  14.550 ms  18.560 ms
 3  ge-1-11-rr01.bolton.ct.hartford.comcast.net (68.87.148.177)  19.749 ms  20.079 ms  20.083 ms
 4  be-115-ar01.woburn.ma.boston.comcast.net (68.85.106.25)  21.603 ms  22.035 ms  22.030 ms
 5  4.68.71.221 (4.68.71.221)  21.022 ms  20.635 ms  23.635 ms
 6  * * *
 7  * * *
 8  212.187.173.54 (212.187.173.54)  93.795 ms  97.363 ms  94.683 ms
 9  ae29.londtw-sbr1.ja.net (146.97.33.9)  97.617 ms  97.928 ms  97.894 ms
10  ae31.lowdss-sbr1.ja.net (146.97.33.29)  103.000 ms  99.365 ms  104.273 ms
11  ae25.warwuw-rbr1.ja.net (146.97.37.158)  106.385 ms  106.715 ms  97.715 ms
12  193.63.208.46 (193.63.208.46)  99.817 ms  99.449 ms  99.280 ms
13  gw-uw-1e.warwick.ac.uk (194.66.251.250)  98.301 ms  97.779 ms  101.323 ms

**Comment by JohnCremona** [comment]
The test_siegel_modular_forms tests all work fine for me running from atkin.

From home: also all work, though just that test took 62s which is rather
absurd.

John

**Comment by AndrewVSutherland** [comment]
Running test_Siegel_link_browse_page works for me as well (using the 
lmfdb version I just pulled from github), but if I copy the same links 
it is using into my browser, they does not work.

I think the problem may be that what is running on the machine serving 
http requests to www.lmfdb.org is not the same as the code currently in 
github.  Which is strange, because the Siegel modular forms code was 
last updated 3 days ago and there have been several pushes since then.

**Comment by JohnCremona** [comment]
Right now all 4 of those links fail at www.lmfdb.org and beta.lmfdb.org which are both running the latest master code.  But from my local server, running the same code, all work fine.  The difference is in the db connection, so that must be at fault.

**Comment by JohnCremona** [comment]
I just pushed master to both beta and prod at warwick to make sure that is
up to date.

**Comment by sanni85** [comment]
For me all the links work now both on www and on beta.

**Comment by AndrewVSutherland** [comment]
And for me, and very snappy too!  I will go ahead and close, but it would be good to understand why/how this happened.  How is it that a push fixed this when there have been several pushes recently?

**Comment by JohnCremona** [comment]
I agree that we really should understand why this suddenly failed.  It also reveals a weakness in our testing system -- running tests locally does not prove that the same URLs wll work on the server.  This has always been true in part, e.g. if your local version was running a different version of something like pymongo or Sage or...

**Comment by jvoight** [comment]
It's not clear that this issue has been resolved.  I'm still having problems ("Internal Server Error"), and they are intermittent.

**Comment by JohnCremona** [comment]
The flasklog shows this on these crashes:

InvalidOperation: MongoReplicaSetClient has been closed

**Comment by edgarcosta** [comment]
We were using MongoReplicaSetClient, which seems to be buggy and not like the way we fetch knowls:
https://gist.github.com/edgarcosta/7b6b4b2243ed8b39ea3f9e6279f0d364

The links started to work as 

However, I would not discard that we were relying on some undefined behaviour. 

Now the servers are running pymongo 3.2.
Which avoids the usage of the deprecated MongoReplicaSetClient, and perhaps introduces new bugs, but I like to believe that 3.2 has less bugs than 2.8.

Further, at 21:51 BST, the mongodb server at warwick stepped down momentarily from being primary, which made some clients freakout.

In summary, you can recreate the setting by using pymongo 2.8 with MongoReplicaSetClient.

Best,
Edgar

**Comment by edgarcosta** [comment]
Here is the reason why it works with pymongo 3.2
http://api.mongodb.com/python/current/api/pymongo/mongo_client.html#pymongo.mongo_client.MongoClient.close

**Comment by AndrewVSutherland** [comment]
So can we close this now?

**Comment by jwj61** [comment]
Yes
